File: python/raft/cliente.py
# Cliente para hablar con el cluster Raft desde Python (lo usa el Servidor de
# Testeo para insertar detecciones). Maneja REDIRECT al lider igual que el cliente
# Java: recalcula el orden priorizando al lider conocido en cada intento.

import socket
import time


class ClienteCluster:
    def __init__(self, nodos):
        # nodos: lista de (id, host, puerto)
        self.nodos = nodos
        self.ultimo_lider = None  # (id, host, puerto)

    def insertar(self, tipo, imagen_ref, fecha_hora, camara):
        cmd = f"{tipo},{imagen_ref},{fecha_hora},{camara}"
        resp = self._pedir(f"NUEVA_DETECCION|{cmd}")
        return resp == "OK"

    def leer_registro(self):
        resp = self._pedir("LEER_REGISTRO")
        if not resp:
            return []
        c = resp.split("|")
        if c[0] != "REGISTRO" or not c[1]:
            return []
        return c[1].split(";")

    def _pedir(self, mensaje):
        for _ in range(len(self.nodos) * 2 + 4):
            orden = []
            if self.ultimo_lider:
                orden.append(self.ultimo_lider)
            for n in self.nodos:
                if not self.ultimo_lider or n[0] != self.ultimo_lider[0]:
                    orden.append(n)

            redirigido = False
            for (nid, host, pto) in orden:
                resp = self._enviar(host, pto, mensaje)
                if resp is None:
                    continue
                c = resp.split("|")
                if c[0] == "REDIRECT":
                    lider = self._por_dir(c[1])
                    if lider and (not self.ultimo_lider or lider[0] != self.ultimo_lider[0]):
                        self.ultimo_lider = lider
                        redirigido = True
                        break
                    continue
                self.ultimo_lider = (nid, host, pto)
                return resp
            if not redirigido:
                time.sleep(0.08)
        return None

    def _por_dir(self, d):
        for n in self.nodos:
            if f"{n[1]}:{n[2]}" == d:
                return n
        return None

    def _enviar(self, host, pto, mensaje):
        try:
            s = socket.create_connection((host, pto), timeout=0.15)
            s.settimeout(0.4)
            f = s.makefile("rwb")
            f.write((mensaje + "\n").encode("utf-8"))
            f.flush()
            resp = f.readline().decode("utf-8").rstrip("\n")
            s.close()
            return resp if resp else None
        except Exception:
            return None


def parsear_nodos(specs):
    out = []
    for spec in specs:
        pid, host, pto = spec.split(":")
        out.append((int(pid), host, int(pto)))
    return out

File: go/internal/raft/cliente.go
package raft

import (
	"bufio"
	"net"
	"strconv"
	"strings"
	"time"
)

// ParsearNodos interpreta specs id:host:puerto (mismo formato que usa
// arrancar-nodo) y arma la lista de pares del cluster.
func ParsearNodos(specs []string) []Par {
	nodos := make([]Par, 0, len(specs))
	for _, spec := range specs {
		p := strings.Split(spec, ":")
		if len(p) != 3 {
			continue
		}
		id, err1 := strconv.Atoi(p[0])
		puerto, err2 := strconv.Atoi(p[2])
		if err1 != nil || err2 != nil {
			continue
		}
		nodos = append(nodos, Par{ID: id, Host: p[1], Puerto: puerto})
	}
	return nodos
}

// ClienteCluster habla con el cluster Raft desde fuera (equivalente a
// java/src/raft/ClienteCluster.java). Sirve para probar el nucleo Go y,
// eventualmente, para cualquier proceso Go que necesite insertar detecciones
// o leer el registro sin ser el nodo mismo. Maneja el REDIRECT: si contacta
// a un seguidor, sigue al lider y reintenta ahi.
type ClienteCluster struct {
	nodos       []Par
	ultimoLider *Par
}

func NuevoClienteCluster(nodos []Par) *ClienteCluster {
	return &ClienteCluster{nodos: nodos}
}

// Insertar agrega una deteccion. Devuelve true si el cluster la acepto.
func (c *ClienteCluster) Insertar(tipo, imagenRef, fechaHora, camara string) bool {
	cmd := strings.Join([]string{tipo, imagenRef, fechaHora, camara}, SepCampoEntrada)
	resp, ok := c.pedir(Unir(NuevaDeteccion, cmd))
	return ok && resp == Ok
}

// LeerRegistro devuelve las detecciones cometidas (cada una es el string
// tipo,imagenRef,fechaHora,camara).
func (c *ClienteCluster) LeerRegistro() []string {
	resp, ok := c.pedir(LeerRegistro)
	if !ok {
		return nil
	}
	campos := Partir(resp)
	if len(campos) < 2 || campos[0] != Registro {
		return nil
	}
	if campos[1] == "" {
		return nil
	}
	return strings.Split(campos[1], SepEntradas)
}

// pedir envia un mensaje probando nodos hasta dar con el lider (siguiendo
// REDIRECT). En cada intento se prioriza al ultimo lider conocido.
func (c *ClienteCluster) pedir(mensaje string) (string, bool) {
	intentosMax := len(c.nodos)*2 + 4
	for intento := 0; intento < intentosMax; intento++ {
		orden := c.ordenConLiderPrimero()

		redirigido := false
		for _, p := range orden {
			resp, ok := enviarACluster(p, mensaje)
			if !ok {
				continue
			}
			campos := Partir(resp)
			if campos[0] == Redirect {
				lider := c.porDireccion(campos[1])
				if lider != nil && (c.ultimoLider == nil || lider.ID != c.ultimoLider.ID) {
					c.ultimoLider = lider
					redirigido = true
					break // reintento de una vez contra el lider correcto
				}
				continue // REDIRECT a "?" o al mismo: sigo probando otros nodos
			}
			p := p
			c.ultimoLider = &p
			return resp, true
		}
		if !redirigido {
			time.Sleep(80 * time.Millisecond) // el cluster puede estar en eleccion
		}
	}
	return "", false
}

func (c *ClienteCluster) ordenConLiderPrimero() []Par {
	orden := make([]Par, 0, len(c.nodos))
	if c.ultimoLider != nil {
		orden = append(orden, *c.ultimoLider)
	}
	for _, p := range c.nodos {
		if c.ultimoLider == nil || p.ID != c.ultimoLider.ID {
			orden = append(orden, p)
		}
	}
	return orden
}

func (c *ClienteCluster) porDireccion(dir string) *Par {
	for _, p := range c.nodos {
		if p.Direccion() == dir {
			pp := p
			return &pp
		}
	}
	return nil
}

func enviarACluster(p Par, mensaje string) (string, bool) {
	conn, err := net.DialTimeout("tcp", p.Direccion(), 150*time.Millisecond)
	if err != nil {
		return "", false
	}
	defer conn.Close()
	_ = conn.SetDeadline(time.Now().Add(400 * time.Millisecond))
	if _, err := conn.Write([]byte(mensaje + "\n")); err != nil {
		return "", false
	}
	r := bufio.NewReader(conn)
	linea, err := r.ReadString('\n')
	if err != nil && linea == "" {
		return "", false
	}
	return trimEOL(linea), true
}
